modules/xfem/src/userobjects/InterfaceMeshCut3DUserObject.C

```diff
@@ -0,0 +1,243 @@
+        Real lenSq1 = line_1 * line_1;
+        Real lenSq2 = line_2 * line_2;
+        Real angle = std::acos(dot / std::sqrt(lenSq1 * lenSq2));
+        normal_at_node += normal_at_node_j * angle;
+        angle_sum += angle;
+      }
+      normal[1 + i] = normal_at_node;
+    }
+
+    for (unsigned int i = 0; i < elem->n_sides(); i++)
+    {
+      std::vector<Point> vertices{elem->node_ref(0), elem->node_ref(1), elem->node_ref(2)};
+
+      Plane elem_plane(vertices[0], vertices[1], vertices[2]);
+      Point normal_at_edge = libMesh::pi * elem_plane.unit_normal(vertices[0]);
+
+      const Elem * neighbor = elem->neighbor_ptr(i);
+
+      if (neighbor != nullptr)
+      {
+        std::vector<Point> vertices{
+            neighbor->node_ref(0), neighbor->node_ref(1), neighbor->node_ref(2)};
+
+        Plane elem_plane(vertices[0], vertices[1], vertices[2]);
+        normal_at_edge += libMesh::pi * elem_plane.unit_normal(vertices[0]);
+      }
+      normal[4 + i] = normal_at_edge;
+    }
+    _pseudo_normal.insert(std::make_pair(elem->id(), normal));
+  }
+}
+
+Point
+InterfaceMeshCut3DUserObject::nodeNormal(const unsigned int & node_id)
+{
+  Point normal(0.0);
+
+  for (const auto & node_neigh_elem_id : _node_to_elem_map[node_id])
+  {
+    const auto & elem = _cutter_mesh->elem_ref(node_neigh_elem_id);
+    Plane elem_plane(elem.node_ref(0), elem.node_ref(1), elem.node_ref(2));
+    normal += elem_plane.unit_normal(elem.node_ref(0));
+  }
+
+  unsigned int num = _node_to_elem_map[node_id].size();
+  return normal / num;
+}
+
+bool
+InterfaceMeshCut3DUserObject::cutElementByGeometry(const Elem * /*elem*/,
+                                                   std::vector<Xfem::CutEdge> & /*cut_edges*/,
+                                                   std::vector<Xfem::CutNode> & /*cut_nodes*/) const
+{
+  mooseError("invalid method for 3D mesh cutting");
+  return false;
+}
+
+bool
+InterfaceMeshCut3DUserObject::cutElementByGeometry(const Elem * elem,
+                                                   std::vector<Xfem::CutFace> & cut_faces) const
+{
+  bool elem_cut = false;
+
+  if (elem->dim() != 3)
+    mooseError("The structural mesh to be cut by a surface mesh must be 3D!");
+
+  for (unsigned int i = 0; i < elem->n_sides(); ++i)
+  {
+    // This returns the lowest-order type of side.
+    std::unique_ptr<const Elem> curr_side = elem->side_ptr(i);
+    if (curr_side->dim() != 2)
+      mooseError("In cutElementByGeometry dimension of side must be 2, but it is ",
+                 curr_side->dim());
+    unsigned int n_edges = curr_side->n_sides();
+
+    std::vector<unsigned int> cut_edges;
+    std::vector<Real> cut_pos;
+
+    for (unsigned int j = 0; j < n_edges; j++)
+    {
+      // This returns the lowest-order type of side.
+      std::unique_ptr<const Elem> curr_edge = curr_side->side_ptr(j);
+      if (curr_edge->type() != EDGE2)
+        mooseError("In cutElementByGeometry face edge must be EDGE2, but type is: ",
```

Comment:
The check in the constructor of this class that they are TRI3 elements should cover this. Maybe put a more detailed error message there, and then just have an assert here.

Reply:
In the constructor, we only check if the cutting mesh is TRI3 while we are checking if the edge of the 3D element is EDGE2.